src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -43,24 +43,14 @@ private enum FrameType : byte
             AppData = 23
         }
 
-        //
-        // This block is used to rule the >>re-handshakes<< that are concurrent with read/write I/O requests.
-        //
-        private const int LockNone = 0;
-        private const int LockWrite = 1;
-        private const int LockHandshake = 2;
-        private const int LockPendingWrite = 3;
-        private const int LockRead = 4;
-        private const int LockPendingRead = 6;
+        private readonly object _handshakeLock = new object();
+        private TaskCompletionSource<bool>? _handshakeWaiter;
```

Comment:
This should be made `volatile`.
```suggestion
        private volatile TaskCompletionSource<bool>? _handshakeWaiter;
```